### PR TITLE
AMLS-4002 | Bank details page split

### PR DESCRIPTION
**Pull request by owen-opencast**
## How Has This Been Tested?

Spec tests, manual testing.
**Still need to flesh out spec tests, but can be tested by QA**.

## Checklist

- [ ] Breaking change (fix or feature that would cause existing functionality to change).
- [x] Build passing.
- [ ] Unit tests have full coverage.
- [ ] Unit test approach and implementation has been reviewed with QA (testers).
- [x] Requires acceptance test run.
- [x] Appropriate labels added.
- [x] RELEASE NOTES ADDED.

### CODE REVIEW

File: app/controllers/bankdetails/BankAccountHasIbanController.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+import scala.concurrent.Future
+
+@Singleton
+class BankAccountHasIbanController @Inject()(
+                                           val dataCacheConnector: DataCacheConnector,
+                                           val authAction: AuthAction,
+                                           val auditConnector: AuditConnector,
+                                           val statusService: StatusService
+                                         ) extends BankDetailsController {
```
Comment (slavvco) [inline app/controllers/bankdetails/BankAccountHasIbanController.scala]: please correct indentation

File: app/controllers/bankdetails/BankAccountUKController.scala [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import scala.concurrent.Future
+
+@Singleton
+class BankAccountUKController @Inject()(
+                                           val dataCacheConnector: DataCacheConnector,
+                                           val authAction: AuthAction,
+                                           val auditConnector: AuditConnector,
+                                           val statusService: StatusService
+                                         ) extends BankDetailsController {
+
```
Comment (slavvco) [inline app/controllers/bankdetails/BankAccountUKController.scala]: correct indentation please

File: app/models/bankdetails/BankAccount.scala [hunk 3]
```diff
@@ -16,137 +16,53 @@
...
 
+  private def changedIsUk(newData: BankAccountIsUk): Boolean =
+    isUk.map(acc => acc.isUk).exists(isUk => isUk != newData.isUk)
 
-  implicit val formRead: Rule[UrlFormEncoded, Account] =
-    From[UrlFormEncoded] { __ =>
-      import jto.validation.forms.Rules._
+  private def changedHasIban(newData: BankAccountHasIban): Boolean =
+    hasIban.map(iban => iban.hasIban).exists(hasIban => hasIban != newData.hasIban)
+}
```
Comment (slavvco) [inline app/models/bankdetails/BankAccount.scala]: lines 42 and 45 I find them a bit hard to read

File: app/controllers/bankdetails/BankAccountHasIbanController.scala [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+              )
+            }.flatMap { _ =>
+                lazy val redirect = if(data.hasIban) {
+                  Redirect(routes.BankAccountIbanController.get(index))
+                } else {
+                  Redirect(routes.BankAccountNonUKController.get(index))
+                }
+                (sendAudit map { _ =>
+                  redirect
+                }) getOrElse redirect
```
Comment (ajchilts) [inline app/controllers/bankdetails/BankAccountHasIbanController.scala]: Possible refactor here to make this easier to follow? Perhaps split out the above into methods that do a single purpose?

File: app/controllers/bankdetails/BankAccountIbanController.scala [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+              if (edit) {
+                Future.successful(Redirect(routes.SummaryController.get(index)))
+              } else {
+                lazy val redirect = Redirect(routes.SummaryController.get(index))
+                (sendAudit map { _ =>
+                  redirect
+                }) getOrElse redirect
+              }
+            }
+        }
```
Comment (ajchilts) [inline app/controllers/bankdetails/BankAccountIbanController.scala]: Possible refactor here to make this easier to follow? Perhaps split out the above into methods that do a single purpose? With particular reference to the update on the object.

File: app/controllers/bankdetails/BankAccountIbanController.scala [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                  StatusConstants.Updated
+                } else {
+                  StatusConstants.Added
+                })
+              )
+            }.flatMap { _ =>
+              if (edit) {
+                Future.successful(Redirect(routes.SummaryController.get(index)))
+              } else {
+                lazy val redirect = Redirect(routes.SummaryController.get(index))
```
Comment (ajchilts) [inline app/controllers/bankdetails/BankAccountIbanController.scala]: Does the redirect need to be in a val here?

File: app/controllers/bankdetails/BankAccountNonUKController.scala [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                })
+              )
+            }.flatMap { _ =>
+              if (edit) {
+                Future.successful(Redirect(routes.SummaryController.get(index)))
+              } else {
+                lazy val redirect = Redirect(routes.SummaryController.get(index))
+                (sendAudit map { _ =>
+                  redirect
+                }) getOrElse redirect
```
Comment (ajchilts) [inline app/controllers/bankdetails/BankAccountNonUKController.scala]: Possible refactor here to make this easier to follow? Perhaps split out the above into methods that do a single purpose? With particular reference to the update on the object.

File: app/controllers/bankdetails/BankAccountUKController.scala [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+              )
+            }.flatMap { _ =>
+              if (edit) {
+                Future.successful(Redirect(routes.SummaryController.get(index)))
+              } else {
+                lazy val redirect = Redirect(routes.SummaryController.get(index))
+                (sendAudit map { _ =>
+                  redirect
+                }) getOrElse redirect
+              }
```
Comment (ajchilts) [inline app/controllers/bankdetails/BankAccountUKController.scala]: As above.

File: app/controllers/bankdetails/BankAccountHasIbanController.scala (hunk 4 above)
Comment (owen-opencast) [reply app/controllers/bankdetails/BankAccountHasIbanController.scala]: I've removed the audit here, as (I think) it should only be triggered on completion of all bank details

File: app/controllers/bankdetails/BankAccountHasIbanController.scala (hunk 1 above)
Comment (owen-opencast) [reply app/controllers/bankdetails/BankAccountHasIbanController.scala]: Done

File: app/controllers/bankdetails/BankAccountUKController.scala (hunk 2 above)
Comment (owen-opencast) [reply app/controllers/bankdetails/BankAccountUKController.scala]: Done

File: app/controllers/bankdetails/BankAccountIbanController.scala (hunk 5 above)
Comment (owen-opencast) [reply app/controllers/bankdetails/BankAccountIbanController.scala]: As you said, perhaps tackle this separately? It's mostly unchanged from the original code.

**Pull request merged**